File: BestTimeToBuyAndSellStock.java
/**
 * Author: Chih-Jye Wang
 * Date  : Dec 17, 2013
 *
 * Say you have an array for which the ith element is the price of a given stock on day i.
 *
 * If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
 *
 * Reference: https://github.com/AnnieKim/LeetCode
 */
public class BestTimeToBuyAndSellStock
{
    public int maxProfit(int[] prices)
    {
        int minPrice = prices[0];
        int max = Integer.MIN_VALUE;
        for(int i = 0; i < prices.length; i++)
        {
            if(prices[i] < minPrice)
                minPrice = prices[i];
            if(prices[i] - minPrice > max)
                max = prices[i] - minPrice;    
        }
        return max;
    }
}

File: ruby/best_time_to_buy_and_sell_stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
    return 0 if !prices || prices.length == 0
    min_price = prices[0]
    res = 0
    for i in 0...prices.length
        min_price = prices[i] if prices[i] < min_price
        res = [res, prices[i] - min_price].max
    end
    return res
end
